### PR TITLE
[dyld] Do not resolve symbols from executables compiled with -fPIE. 

### LINKED ISSUE #7366: ACLiC compilation confuses compiled binaries with shared objects, breaking compilation in some cases
- [X] Checked for duplicates

### Describe the bug

Building ROOT on Arch Linux and then running `ctest -R "(dataframe|datasource)" -- -j8` results in several test failures.
There are various possible failure modes -- the command above typically results in these kind of errors:

```
1369: Processing /home/jalopezg/CERN/roottest/root/dataframe/test_snapshot.C+...
1369: Info in <TUnixSystem::ACLiC>: creating shared library /home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C.so
1369: /usr/bin/ld: /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc: _ZSt4cout: invalid version 6 (max 0)
1369: /usr/bin/ld: /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc: error adding symbols: bad value
1369: collect2: error: ld returned 1 exit status
1369: Error in <ACLiC>: Executing 'cd "/home/jalopezg/CERN/build/roottest/root/dataframe" ; c++ -fPIC -c -O3 -DNDEBUG -std=c++14 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe -W -Woverloaded-virtual -fsigned-char -pthread  -I$ROOTSYS/include -I/home/jalopezg/CERN/build/roottest/root/dataframe -I"/home/jalopezg/CERN/build/etc/" -I"/home/jalopezg/CERN/build/etc//cling" -I"/home/jalopezg/CERN/build/include/" -I"/home/jalopezg/CERN/build/include" -I"/home/jalopezg/CERN/build/roottest/root/dataframe"   -D__ACLIC__ "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C_ACLiC_dict.cxx" ; c++ -O3 -DNDEBUG "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C_ACLiC_dict.o" -shared   "/usr/lib/libdl.so" "/usr/lib/libc.so" "/usr/lib/libgcc_s.so" "/usr/lib/libm.so" "/usr/lib/libstdc++.so" "/home/jalopezg/CERN/build/lib/libRint.so" "/home/jalopezg/CERN/build/lib/libCore.so" "/usr/lib/libpthread.so.0" "/usr/lib/libpcre.so.1" "/usr/lib/liblzma.so.5" "/usr/lib/liblz4.so.1" "/usr/lib/libz.so.1" "/usr/lib/libzstd.so.1" "/usr/lib/libnss_files.so.2" "/home/jalopezg/CERN/build/lib/libRIO.so" "/home/jalopezg/CERN/build/lib/libThread.so" "/home/jalopezg/CERN/build/lib/libCling.so" "/usr/lib/librt.so.1" "/usr/lib/libncursesw.so.6" "/usr/lib/ld-2.33.so" "/home/jalopezg/CERN/build/lib/libMathCore.so" "/home/jalopezg/CERN/build/lib/libImt.so" "/home/jalopezg/CERN/build/lib/libMultiProc.so" "/home/jalopezg/CERN/build/lib/libNet.so" "/usr/lib/libtbb.so.2" "/usr/lib/libssl.so.1.1" "/usr/lib/libcrypto.so.1.1" "/usr/lib/libnss_systemd.so.2" -o "/home/jalopezg/CERN/build/roottest/root/dataframe/test_snapshot_C.so" /home/jalopezg/CERN/build/lib/libROOTDataFrame.so /home/jalopezg/CERN/build/lib/libTreePlayer.so /home/jalopezg/CERN/build/lib/libTree.so /home/jalopezg/CERN/build/roottest/root/dataframe/test_readFcc /home/jalopezg/CERN/build/roottest/root/dataframe/par' failed!
1369: terminate called after throwing an instance of 'std::logic_error'
1369:   what():  basic_string::_M_construct null not valid
```

I did not manage to reproduce the problem outside of ctest and without running multiple tests concurrently -- it also seems that one of the tests "has" to be `roottest-root-dataframe-test_snapshot_manytasks`.

### To Reproduce

The following repository provides a Dockerfile to reproduce the issue in a container https://gitlab.cern.ch/eguiraud/arch_aclic_bug .

### Setup

ROOT master (but 6.22 also has issues) and arch linux.

I tried building from source (6.22 and master) as well as installing from the system repos -- failure modes are different but none can run all tests without problems.

Gentoo seems to also be affected.

### PR DESCRIPTION
**Pull request by vgvassilev**
Executables that are compiled with fPIE means they are compiled in a position independent manner and are almost indistinguishable from the shared objects. A reasonably reliable way to find if this was a `pie executable` is to check the `DF_1_PIE` in the dynamic section of ELF.

The pseudo-code is:
```
    if DT_FLAGS_1 dynamic section entry is present
      if DF_1_PIE is set in DT_FLAGS_1:
        print pie executable
      else
        print shared object
```
    
    See https://stackoverflow.com/questions/34519521/why-does-gcc-create-a-shared-object-instead-of-an-executable-binary-according-to/34522357#34522357

Fixes #7366

Patch by Alexander Penev (@alexander-penev)

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
Any chance to have a test?

**Comment by Axel-Naumann** [comment]
And: congrats for solving this!

**Comment by vgvassilev** [comment]
@Axel-Naumann ping.

File: interpreter/cling/test/DynamicLibraryManager/call_pie.c [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+// RUN: true
+// Used as executable/library source by pie.C, etc.
+CLING_EXPORT int cling_testlibrary_function() {
+  return 42;
+}
+
+int main() {
+  return 0;
+}
```
Comment (Axel-Naumann) [inline interpreter/cling/test/DynamicLibraryManager/call_pie.c:17]: Newline at EOF?
```suggestion
}
```

**Review by Axel-Naumann — APPROVED** [review]
Sorry for the late review - LGTM, thanks! (and see missing newline at EOF)

**Pull request merged**